File: client-libs/typescript/src/utils/Validators.ts
/** DAG Validation Utilities

Provides validation utilities for DAG structures before submission to the API.
*/

import { BaseDAGBuilder } from "../builders/BaseDAGBuilder";
import type { DAGDefinition, DAGValidationResult } from "../types/dag.types";

export class DAGValidator {
  /**
   * Validate DAG structure
   */
  static validateDAG(dag: DAGDefinition): DAGValidationResult {
    return BaseDAGBuilder.validateDAG(dag);
  }

  /**
   * Validate all UUIDs in DAG are unique
   */
  static validateUUIDUniqueness(dag: DAGDefinition): DAGValidationResult {
    const uuids = dag.map((node) => node.uuid);
    const uniqueUuids = new Set(uuids);

    if (uuids.length !== uniqueUuids.size) {
      const duplicates = uuids.filter((uuid, index) => uuids.indexOf(uuid) !== index);
      return {
        isValid: false,
        error: `Duplicate UUIDs found: ${[...new Set(duplicates)].join(", ")}`,
      };
    }

    return { isValid: true };
  }

  /**
   * Validate all edge targets exist in DAG
   */
  static validateEdges(dag: DAGDefinition): DAGValidationResult {
    const allUuids = new Set(dag.map((node) => node.uuid));

    for (const node of dag) {
      for (const edgeTarget of node.edges || []) {
        if (!allUuids.has(edgeTarget)) {
          return {
            isValid: false,
            error: `Edge target '${edgeTarget}' from node '${node.uuid}' not found in DAG`,
          };
        }
      }
    }

    return { isValid: true };
  }

  /**
   * Validate all required fields are present
   */
  static validateRequiredFields(dag: DAGDefinition): DAGValidationResult {
    const requiredFields: (keyof typeof dag[0])[] = ["uuid", "job_title", "job_type", "data"];

    for (const node of dag) {
      for (const field of requiredFields) {
        if (!(field in node)) {
          return {
            isValid: false,
            error: `Required field '${field}' missing in node ${node.uuid || "unknown"}`,
          };
        }
      }
    }

    return { isValid: true };
  }

  /**
   * Validate node count is within limits
   */
  static validateNodeCount(dag: DAGDefinition): DAGValidationResult {
    if (!dag || dag.length === 0) {
      return { isValid: false, error: "DAG cannot be empty" };
    }

    if (dag.length > BaseDAGBuilder.MAX_NODES_PER_REQUEST) {
      return {
        isValid: false,
        error: `DAG exceeds maximum nodes limit (${BaseDAGBuilder.MAX_NODES_PER_REQUEST}): ${dag.length} nodes`,
      };
    }

    return { isValid: true };
  }
}

File: client-libs/python/utils/validators.py
"""DAG Validation Utilities

Provides validation utilities for DAG structures before submission to the API.
"""

from typing import Any, Dict, List, Optional

from ..dag_builders.base import BaseDAGBuilder


class DAGValidator:
    """Validator for DAG structures"""

    @staticmethod
    def validate_dag(dag: List[Dict[str, Any]]) -> tuple[bool, Optional[str]]:
        """Validate DAG structure
        
        This is a wrapper around BaseDAGBuilder.validate_dag for consistency.
        
        Args:
            dag: List of job nodes
            
        Returns:
            Tuple of (is_valid, error_message)
        """
        return BaseDAGBuilder.validate_dag(dag)

    @staticmethod
    def validate_uuid_uniqueness(dag: List[Dict[str, Any]]) -> tuple[bool, Optional[str]]:
        """Validate all UUIDs in DAG are unique
        
        Args:
            dag: List of job nodes
            
        Returns:
            Tuple of (is_valid, error_message)
        """
        uuids = [node.get("uuid") for node in dag]
        unique_uuids = set(uuids)
        
        if len(uuids) != len(unique_uuids):
            duplicates = [uuid for uuid in uuids if uuids.count(uuid) > 1]
            return False, f"Duplicate UUIDs found: {set(duplicates)}"
        
        return True, None

    @staticmethod
    def validate_edges(dag: List[Dict[str, Any]]) -> tuple[bool, Optional[str]]:
        """Validate all edge targets exist in DAG
        
        Args:
            dag: List of job nodes
            
        Returns:
            Tuple of (is_valid, error_message)
        """
        all_uuids = {node.get("uuid") for node in dag}
        
        for node in dag:
            edges = node.get("edges", [])
            for edge_target in edges:
                if edge_target not in all_uuids:
                    return (
                        False,
                        f"Edge target '{edge_target}' from node '{node.get('uuid')}' not found in DAG",
                    )
        
        return True, None

    @staticmethod
    def validate_required_fields(dag: List[Dict[str, Any]]) -> tuple[bool, Optional[str]]:
        """Validate all required fields are present
        
        Args:
            dag: List of job nodes
            
        Returns:
            Tuple of (is_valid, error_message)
        """
        required_fields = ["uuid", "job_title", "job_type", "data"]
        
        for node in dag:
            for field in required_fields:
                if field not in node:
                    return (
                        False,
                        f"Required field '{field}' missing in node {node.get('uuid', 'unknown')}",
                    )
        
        return True, None

    @staticmethod
    def validate_node_count(dag: List[Dict[str, Any]]) -> tuple[bool, Optional[str]]:
        """Validate node count is within limits
        
        Args:
            dag: List of job nodes
            
        Returns:
            Tuple of (is_valid, error_message)
        """
        if not dag:
            return False, "DAG cannot be empty"
        
        if len(dag) > BaseDAGBuilder.MAX_NODES_PER_REQUEST:
            return (
                False,
                f"DAG exceeds maximum nodes limit ({BaseDAGBuilder.MAX_NODES_PER_REQUEST}): {len(dag)} nodes",
            )
        
        return True, None
